operator/cmd/metrics/adapter/provider/provider.go

```diff
@@ -0,0 +1,178 @@
+func (p *keptnMetricsProvider) ListAllMetrics() []provider.CustomMetricInfo {
+	return p.cache.List()
+}
+
+// GetMetricByName retrieves a metric based on its name.
+// Used for requests such as e.g. /apis/custom.metrics.k8s.io/v1beta2/namespaces/keptn-lifecycle-toolkit/keptnmetrics.metrics.sh/keptnmetric-sample/keptnmetric-sample
+func (p *keptnMetricsProvider) GetMetricByName(ctx context.Context, name types.NamespacedName, info provider.CustomMetricInfo, metricSelector labels.Selector) (*custom_metrics.MetricValue, error) {
+	klog.InfoS("GetMetricByName()", "name", name, "metricSelector", metricSelector, "context", ctx)
+	val, err := p.cache.Get(name.Name)
+	if err != nil {
+		if errors.Is(err, ErrMetricNotFound) {
+			return nil, provider.NewMetricNotFoundForSelectorError(info.GroupResource, info.Metric, name.Name, metricSelector)
+		}
+		return nil, &apierr.StatusError{ErrStatus: metav1.Status{
+			Status:  metav1.StatusFailure,
+			Code:    int32(http.StatusInternalServerError),
+			Reason:  metav1.StatusReasonInternalError,
+			Message: err.Error(),
+		}}
+	}
+	return &val.Value, nil
+}
+
+// GetMetricBySelector retrieves a list of metrics based on the given selectors.
+// Used for requests such as e.g. /apis/custom.metrics.k8s.io/v1beta2/namespaces/keptn-lifecycle-toolkit/keptnmetrics.metrics.sh/*/*?labelSelector=<key>%3D<value>
+func (p *keptnMetricsProvider) GetMetricBySelector(ctx context.Context, _ string, selector labels.Selector, _ provider.CustomMetricInfo, metricSelector labels.Selector) (*custom_metrics.MetricValueList, error) {
+	klog.InfoS("GetMetricBySelector()", "selector", selector, "metricSelector", metricSelector, "context", ctx)
+
+	metricValues := p.cache.GetValuesByLabel(selector)
+
+	res := make([]custom_metrics.MetricValue, len(metricValues))
+	i := 0
+	for _, metricValue := range metricValues {
+		res[i] = metricValue.Value
+		i++
+	}
+
+	return &custom_metrics.MetricValueList{
+		Items: res,
+	}, nil
+}
+
+func (p *keptnMetricsProvider) watchMetrics(ctx context.Context) error {
+	factory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(p.client, 0, kltNamespace, nil)
+
+	informer := factory.ForResource(keptnMetricGroupVersionResource).Informer()
+
+	handlers := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			klog.InfoS("AddFunc", "obj", obj)
+			p.updateMetric(obj)
+		},
+		UpdateFunc: func(oldObj, obj interface{}) {
+			klog.InfoS("UpdateFunc", "obj", obj)
+			p.updateMetric(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			klog.InfoS("DeleteFunc", "obj", obj)
+			unstructuredKeptnMetric := obj.(*unstructured.Unstructured)
+			p.cache.Delete(unstructuredKeptnMetric.GetName())
+		},
+	}
+	if _, err := informer.AddEventHandler(handlers); err != nil {
+		return err
+	}
+	go func() {
+		informer.Run(ctx.Done())
+	}()
+	return nil
+}
+
+func (p *keptnMetricsProvider) updateMetric(obj interface{}) {
+	unstructuredKeptnMetric := obj.(*unstructured.Unstructured)
```

Comment:
I think we could use the proper type here

Reply:
I tried it out, but the factory for creating an informer on structured objects is not aware of our CRDs. 
In the `updateMetric` function i tried to use `runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), &keptnMetric)` to cast to a KeptnMetric struct but this was not working unfortunately with the following error being thrown during the conversion:

```
err="unrecognized type: uint"
```

Therefore, I would suggest to keep the current approach